Source/NSFontCollection.m

```diff
@@ -0,0 +1,341 @@
+- (NSArray *) matchingDescriptorsForFamily: (NSString *)family options: (NSDictionary *)options
+{
+  NSMutableArray *r = [NSMutableArray arrayWithCapacity: 50];
+  NSArray *a = [self matchingDescriptorsWithOptions: options];
+  NSEnumerator *en = [a objectEnumerator];
+  id o = nil;
+
+  while((o = [en nextObject]) != nil)
+    {
+      if ([[o familyName] isEqualToString: family])
+        {
+          [r addObject: o];
+        }    
+    }
+  
+  return [a copy];
+}
+
+- (instancetype) copyWithZone: (NSZone *)zone
+{
+  NSFontCollection *fc = [[NSFontCollection allocWithZone: zone] init];
+
+  ASSIGNCOPY(fc->_fonts, _fonts);
+  ASSIGNCOPY(fc->_queryDescriptors, _queryDescriptors);
+  ASSIGNCOPY(fc->_exclusionDescriptors, _exclusionDescriptors);
+  ASSIGNCOPY(fc->_queryAttributes, _queryAttributes);
+
+  return fc;
+}
+
+- (instancetype) mutableCopyWithZone: (NSZone *)zone
+{
+  NSMutableFontCollection *fc = [[NSMutableFontCollection allocWithZone: zone] init];
+
+  [fc _setFonts: _fonts];
+  [fc setQueryDescriptors: _queryDescriptors];
+  [fc setExclusionDescriptors: _exclusionDescriptors];
+  [fc _setQueryAttributes: _queryAttributes];
+
+  return fc;
+}
+
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  self = [super init];
+  if (self != nil)
+    {
+      if ([coder allowsKeyedCoding])
+        {
+        }
+    }
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+}
+  
+@end
+ 
+
+@implementation NSMutableFontCollection 
+
++ (NSMutableFontCollection *) fontCollectionWithDescriptors: (NSArray *)queryDescriptors
+{
+  NSMutableFontCollection *fc = [[NSMutableFontCollection alloc] init];
+  ASSIGNCOPY(fc->_queryDescriptors, queryDescriptors);
+  return fc;
+}
+
++ (NSMutableFontCollection *) fontCollectionWithAllAvailableDescriptors
+{
+  return [self fontCollectionWithDescriptors:
+                 [[GSFontEnumerator sharedEnumerator] availableFontDescriptors]];
+}
+
++ (NSMutableFontCollection *) fontCollectionWithLocale: (NSLocale *)locale
+{
+  return [self fontCollectionWithAllAvailableDescriptors];
+}
+
++ (NSMutableFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+{
+  return [__sharedFontCollections objectForKey: name];
+}
+
++ (NSMutableFontCollection *) fontCollectionWithName: (NSFontCollectionName)name
+                                          visibility: (NSFontCollectionVisibility)visibility
+{
+  return [__sharedFontCollections objectForKey: name];
+}
+
+- (NSArray *) queryDescriptors
+{
+  return _queryDescriptors;
+}
+
+- (void) setQueryDescriptors: (NSArray *)queryDescriptors
+{
+  ASSIGN(_queryDescriptors, [queryDescriptors mutableCopy]);
+}
+
+- (NSArray *) exclusionDescriptors
+{
+  return _exclusionDescriptors;
+}
+
+- (void) setExclusionDescriptors: (NSArray *)exclusionDescriptors
+{
+  ASSIGN(_exclusionDescriptors, [exclusionDescriptors mutableCopy]);
+}
+
+- (void)addQueryForDescriptors: (NSArray *)descriptors
+{
+  [_queryDescriptors addObjectsFromArray: descriptors];
```

Comment:
The documentation is a bit unclear here, but I would expect that this method also removes the descriptors from _exclusionDescriptors. While the next method does the opposite. This should be fairly easy to test on Cocoa.